src/kill-parallel.c

```diff
@@ -0,0 +1,261 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <string.h>
+#include <signal.h>
+#include <poll.h>
+#include <Rinternals.h>
+#include "posix.h"
+#include "ps-internal.h"
+
+// Should that be a global option shared by processx and ps?
+#define PS_INTERRUPT_INTERVAL 200
+
+static struct sigaction old_sig_handler;
+static int kill_parallel_pipe = -1;
+
+static void kill_parallel_sigchld_callback(int sig, siginfo_t *info, void *ctx);
+static void kill_parallel_setup_sigchld(struct sigaction *prev);
+static void kill_parallel_remove_sigchld(struct sigaction *prev);
+static void block_sigchld_save(sigset_t *old);
+static void procmask_set(sigset_t *set);
+static void unblock_sigchld(void);
+
+SEXP ps__kill_parallel(SEXP ps, SEXP ffi_grace) {
```

Comment:
This is a variant of the wait routine in processx that supports multiple processes.